File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.py
from collections import Counter

def make_multiset(s):
    s = s.lower()
    multiset = []
    
    for i in range(len(s) - 1):
        a, b = s[i], s[i+1]
        if a.isalpha() and b.isalpha():
            multiset.append(a + b)
    return multiset

def solution(str1, str2):
    # A = [fr, ra, an, nc, ce]
    A = make_multiset(str1)
    B = make_multiset(str2)
    
    # counterA = {'fr': 1, 'ra': 1, ...}
    counterA = Counter(A)
    counterB = Counter(B)
    
    # &: 교집합, 양쪽 모두에 존재할 경우 더 적은 쪽의 개수
    # inter = {'fr': 1, 'nc': 1}
    # Counter는 딕셔너리, 리스트 모두 지원하고,
    # dict 상속받은 특수 클래스라
    # &, | 연산자는 원래 리스트만 가능하지만
    # 오버라이딩해서 이것도 계산 됨
    inter = counterA & counterB
    # |: 합집합, 양쪽 모두에 존재할 경우 더 많은 쪽의 개수
    union = counterA | counterB
    
    inter_cnt = sum(inter.values())
    union_cnt = sum(union.values())
    
    if union_cnt == 0:
        return 65536
    
    return int((inter_cnt / union_cnt) * 65536)

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.js
function makeMultiset(s) {
    s = s.toLowerCase();
    const multiset = [];

    for (let i = 0; i < s.length - 1; i++) {
        const a = s[i];
        const b = s[i + 1];
        
        if (/[a-z]/.test(a) && /[a-z]/.test(b)) {
            multiset.push(a + b);
        }
    }
    
    return multiset;
}

function countMap(arr) {
    const map = new Map();
    for (const item of arr) {
        map.set(item, (map.get(item) || 0 ) + 1);
    }
    return map;
}

function getInter(mapA, mapB) {
    const inter = new Map();
    for (const [key, valA] of mapA) {
        if (mapB.has(key)) {
            const valB = mapB.get(key);
            inter.set(key, Math.min(valA, valB));
        }
    }
    return inter;
}

function getUnion(mapA, mapB) {
    const union = new Map(mapA);
    for (const [key, valB] of mapB) {
        const valA = mapA.get(key) || 0;
        union.set(key, Math.max(valA, valB));
    }
    return union;
}

function sumValues(map) {
    let total = 0;
    for (const val of map.values()) {
        total += val;
    }
    return total;
}


function solution(str1, str2) {
    const A = makeMultiset(str1);
    const B = makeMultiset(str2);
    
    const counterA = countMap(A);
    const counterB = countMap(B);
    
    const inter = getInter(counterA, counterB);
    const union = getUnion(counterA, counterB);
    
    const interCnt = sumValues(inter);
    const unionCnt = sumValues(union);
    
    if (unionCnt === 0) return 65536;
    
    return Math.floor((interCnt / unionCnt) * 65536);
}